stdlib/src/os/path/path.mojo

```diff
@@ -390,3 +391,154 @@ def split[PathLike: os.PathLike, //](path: PathLike) -> (String, String):
 #         paths_str.append(cur_path[].__fspath__())
 
 #     return join(path.__fspath__(), *paths_str)
+
+
+# ===----------------------------------------------------------------------=== #
+# expandvars
+# ===----------------------------------------------------------------------=== #
+
+
+fn _is_shell_special_variable(byte: Byte) -> Bool:
+    """Checks if `$` + `byte` identifies a special shell variable, such as `$@`.
+
+    Args:
+        byte: The byte to check.
+
+    Returns:
+        True if the byte is a special shell variable and False otherwise.
+    """
+    alias shell_variables = InlineArray[Int, 17](
+        ord("*"),
+        ord("#"),
+        ord("$"),
+        ord("@"),
+        ord("!"),
+        ord("?"),
+        ord("-"),
+        ord("0"),
+        ord("1"),
+        ord("2"),
+        ord("3"),
+        ord("4"),
+        ord("5"),
+        ord("6"),
+        ord("7"),
+        ord("8"),
+        ord("9"),
+    )
+    return int(byte) in shell_variables
+
+
+fn _is_alphanumeric(byte: Byte) -> Bool:
+    """Checks if `byte` is an ASCII letter, number, or underscore.
+
+    Args:
+        byte: The byte to check.
+
+    Returns:
+        True if the byte is an ASCII letter, number, or underscore and False otherwise.
+    """
+    var b = int(byte)
+    return (
+        b == ord("_")
+        or ord("0") <= b
+        and b <= ord("9")
+        or ord("a") <= b
+        and b <= ord("z")
+        or ord("A") <= b
+        and b <= ord("Z")
+    )
+
+
+fn _parse_variable_name(bytes: Span[Byte]) -> Tuple[String, Int]:
```

Comment:
That's a good point, we don't need to allocate a new string *every* time. Seems like it's a relatively small change as well, I'll just need to adjust the `${}` or `${` case since it was returning a new hardcoded string. I'll take a look into making this change alongside the suggestions immediately below this.

Reply:
I've combined the cases into one branch, and I switched over to returning an immutable StringSlice.